hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerStateManagerImpl.java

```diff
@@ -260,11 +260,23 @@ private void initialize() throws IOException {
   }
 
   private Map<LifeCycleEvent, CheckedConsumer<ContainerInfo, IOException>>
-      getContainerStateChangeActions() {
+      getContainerStateChangeActions(final Configuration conf) {
     final Map<LifeCycleEvent, CheckedConsumer<ContainerInfo, IOException>>
         actions = new EnumMap<>(LifeCycleEvent.class);
     actions.put(FINALIZE, info -> pipelineManager
         .removeContainerFromPipeline(info.getPipelineID(), info.containerID()));
+
+    boolean shouldRemoveDeletedContainer = conf.getBoolean(
+        ScmConfigKeys.OZONE_SCM_REMOVE_DELETED_CONTAINER_ENABLED,
+        ScmConfigKeys.OZONE_SCM_REMOVE_DELETED_CONTAINER_ENABLED_DEFAULT);
+    if (shouldRemoveDeletedContainer) {
+      // when a CLEANUP event if fired , the state of the container will be
+      // changed to DELETED. at this time, we should remove the reference
+      // of the container from scm. this happens immediately, thus DELETED
+      // is a transient state and hardly be captured.
+      actions.put(CLEANUP, info ->
+          removeContainer(info.containerID().getProtobuf()));
```

Comment:
I am not sure if this is logged anywhere else, but could we ensure a lot message is written at INFO level saying container xxx has been deleted as it is empty.